src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp

```diff
@@ -3039,6 +3039,30 @@ void KviIrcServerParser::parseNumericSaslFail(KviIrcMessage * msg)
 		pOut->output(KVI_OUT_SERVERINFO, __tr2qs("SASL authentication error: %Q"), &szParam);
 	}
 
+	// Handle fallback if possible for SASL auth failure or dump if user mandates SASL
+	// and no fallback is available
+	if(msg->numeric() == 904)
+	{
+		if(msg->connection()->stateData()->sentSaslMethod() == QStringLiteral("EXTERNAL"))
+		{
+			if(!msg->connection()->target()->server()->saslNick().isEmpty() && !msg->connection()->target()->server()->saslPass().isEmpty())
```

Comment:
I'd probably refactor this to not do the same several times...

Reply:
Do what multiple times?

Reply:
Have several pieces of code which check whether PLAIN is valid.
If one will need to be changed later, the other one will be forgotten.